docs/proposals/TrafficMirror.md

```diff
@@ -0,0 +1,92 @@
+# Traffic Mirroring
+
+## Context
+
+Traffic Mirroring is a useful feature which sends a copy of live traffic to the mirrored service which will help in testing a service with real production traffic.The difference is that the reply from the mirrored service is dropped by the envoy proxy and not returned to the caller service.
+
+        a                      b                     c
+    |-----------|             |-----------|    mirror   |------------|
+    | service A | ----------> | Service B | ----------> |  Service C |
+    |---------- |             |-----------|             |------------|
+## Use cases
+
+1. **As a** service owner
+
+   **I want to** illuminate the service under test with production traffic
+
+   **so that** it help reveal bugs that went untested during functional tests and experiment with new features.
+
+2. **As a** service owner
+
+   **I want to** test the service under test with production traffic
+
+   **so that** I can avoid canary deployment.
+## Envoy support
+
+Envoy allows to configure `route.request_mirror_policies` on each Route:
+```json
+{
+  "request_mirror_policies": []
+}
+```
+Where `request_mirror_policies` is an array of elements:
+```json
+{
+  "cluster": "...",
+  "runtime_fraction": "{...}",
+  "trace_sampled": "{...}"
+}
+```
+So traffic mirror in envoy accepts a list of cluster name and the amount of traffic to mirror.[Traffic Mirror](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto.html#envoy-v3-api-msg-config-route-v3-routeaction-requestmirrorpolicy)
+## Proposed configuration model
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: TrafficRoute
+mesh: default
+metadata:
+  name: route-default
+spec:
+  sources:
+    - match:
+        kuma.io/service: sleep_mirror-demo_svc_80
+        app: sleep
+  destinations:
+    - match:
+        kuma.io/service: httpbin_mirror-demo_svc_8080
+  conf:
+    http:
+      - match:
+          path:
+            prefix: "/"
+        split:
+          - weight: 100
+            destination:
+              kuma.io/service: httpbin_mirror-demo_svc_8080
+              version: v1
+          - weight: 0
+            destination:
+              kuma.io/service: httpbin_mirror-demo_svc_8080
+              version: v2
+        mirror:
```

Comment:
If I'm reading Envoy docs right, then this mirror policy should be a part of an existing Envoy route. New envoy routes are defined by `split` section. If `mirror` is a separate section that means we either should create a new route (which probably won't work according to the docs) or guarantee that route is already defined in the `split` section. 

**Example:**
if you define a mirror:
```yaml
mirror:
  destination:
    kuma.io/service: backend
    version: v2
```
then apparently you should guarantee that there is a `split`:
```yaml
- weight: 0
  destination:
  kuma.io/service: httpbin_mirror-demo_svc_8080
  version: v2
```
and that's probably the reason why you added this `split` here with zero weight. To be honest I don't think it's a good way of defining mirrors. 

As an alternative, we can add an option `mirrorTo` or something like this right in the `split`:
```yaml
split:
- weight: 90
  destination:
    kuma.io/service: httpbin_mirror-demo_svc_8080
    version: v1
- weight: 10
  mirrorTo: my_mirror_cluster
  destination:
    kuma.io/service: httpbin_mirror-demo_svc_8080
    version: v2
```
But that's might look overcomplicated and probably we should keep `split` as simple as possible. 

Another option would be to design a separate TrafficMirror policy, but in order to do that, we have to create a proposal on how to match policies for routes (today `source` and `destination` selector able to match only data plane proxies).

It would be great to hear other team members' opinions @kumahq/kuma-maintainers.